File: frontend/src/api/reports.ts
/**
 * 报告管理API客户端
 *
 * 提供报告生成、列表查询、删除、下载功能
 */

// 开发模式和生产模式都使用相对路径
// 开发模式：通过Vite proxy转发
// 生产模式：通过nginx反向代理转发
const API_BASE_URL = '';

/**
 * 获取当前用户ID
 */
export function getCurrentUserId(): string {
  const userStr = localStorage.getItem('librechat_user');

  if (!userStr) {
    throw new Error('未登录：请先访问 LibreChat (/librechat) 登录后再使用本系统');
  }

  try {
    const user = JSON.parse(userStr);
    const userId = user?.id;  // 强制使用 ObjectId

    if (!userId) {
      throw new Error('用户信息不完整：无法获取用户 ID 或邮箱');
    }

    return userId;
  } catch (e) {
    if (e instanceof Error && e.message.includes('用户信息不完整')) {
      throw e;
    }
    throw new Error('用户信息解析失败：localStorage 数据格式错误');
  }
}

/**
 * 通用请求函数
 */
async function apiRequest<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${endpoint}`, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      'X-User-Id': getCurrentUserId(),
      ...options.headers
    }
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    // 处理 FastAPI 的验证错误（detail 可能是数组或对象）
    const errorMessage = typeof errorData.detail === 'string'
      ? errorData.detail
      : JSON.stringify(errorData.detail || errorData, null, 2);
    throw new Error(errorMessage || `API请求失败: ${response.status}`);
  }

  return response.json();
}

// ========== 类型定义 ==========

export interface ReportMeta {
  report_id: string;
  report_name: string;
  model_type: string;
  data_source: string;
  created_at: number;
  metrics: {
    aic?: number;
    bic?: number;
    r2?: number;
  };
}

export interface ReportGenerateParams {
  report_id: string;
  preprocessing_record: any;
  test_results: any;
  descriptive_stats: any;
  data_source_info: {
    name: string;
    source: string;
    fetch_time?: string;
    date_range?: string;
    original_count?: number;
  };
  model_type: string;
}

export interface ReportGenerateResult {
  success: boolean;
  report_id: string;
  report_path: string;
  message: string;
}

export interface ReportListResult {
  success: boolean;
  reports: ReportMeta[];
}

export interface DeleteReportResult {
  success: boolean;
  message: string;
}

export interface ReportDetailsResult {
  success: boolean;
  report_id: string;
  meta: any;
  model_result: any;
}

// ========== API函数 ==========

/**
 * 生成报告（在已有的report_id目录中）
 * @param params - 报告生成参数
 * @returns 生成结果
 */
export async function generateReport(params: ReportGenerateParams): Promise<ReportGenerateResult> {
  // 调试日志：输出实际发送的参数
  console.log('[Debug] generateReport 请求参数:', JSON.stringify(params, null, 2));

  return apiRequest<ReportGenerateResult>('/api/reports/generate', {
    method: 'POST',
    body: JSON.stringify(params)
  });
}

/**
 * 获取报告列表
 * @returns 报告列表
 */
export async function listReports(): Promise<ReportListResult> {
  return apiRequest<ReportListResult>('/api/reports/list', {
    method: 'GET'
  });
}

/**
 * 软删除报告
 * @param reportId - 报告ID
 * @returns 删除结果
 */
export async function deleteReport(reportId: string): Promise<DeleteReportResult> {
  return apiRequest<DeleteReportResult>(`/api/reports/${reportId}`, {
    method: 'DELETE'
  });
}

/**
 * 获取报告详细信息（包含完整模型结果）
 * @param reportId - 报告ID
 * @returns 报告详细信息
 */
export async function getReportDetails(reportId: string): Promise<ReportDetailsResult> {
  return apiRequest<ReportDetailsResult>(`/api/reports/${reportId}/details`, {
    method: 'GET'
  });
}

/**
 * 下载报告DOCX文件
 * @param reportId - 报告ID
 * @returns 下载URL（浏览器会自动下载）
 */
export function getReportDownloadUrl(reportId: string): string {
  return `${API_BASE_URL}/api/reports/download/${reportId}/report.docx`;
}

/**
 * 下载报告DOCX文件（使用fetch支持header传递）
 * @param reportId - 报告ID
 */
export async function downloadReportFile(reportId: string): Promise<void> {
  try {
    const url = getReportDownloadUrl(reportId);
    const response = await fetch(url, {
      headers: {
        'X-User-Id': getCurrentUserId()
      }
    });

    if (!response.ok) {
      throw new Error(`下载失败: ${response.status}`);
    }

    const blob = await response.blob();
    const downloadUrl = URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = downloadUrl;
    link.download = `${reportId}.docx`;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    URL.revokeObjectURL(downloadUrl);
  } catch (error) {
    console.error('下载报告失败:', error);
    throw error;
  }
}

File: stock-mcp/src/server/api/routes/reports.py
"""报告生成API路由"""

from fastapi import APIRouter, HTTPException
from fastapi.responses import FileResponse
from pydantic import BaseModel
from typing import Optional, Dict, Any, List
import os
import json
from src.server.domain.services.report_generator import ReportGenerator
from src.server.utils.request_context import get_current_user_id
from src.server.utils.logger import logger

router = APIRouter(prefix="/api/reports", tags=["reports"])

class ReportRequest(BaseModel):
    """报告生成请求"""
    report_id: str  # 报告ID（与模型构建时生成的report_id一致）
    preprocessing_record: Dict[str, Any]
    test_results: Dict[str, Any]
    descriptive_stats: Dict[str, Any]  # ✅ 支持单变量Dict[str, float]和多变量Dict[str, Dict[str, float]]
    data_source_info: Dict[str, Any]  # ✅ 修改：允许数字类型（original_count），与其他字段保持一致
    model_type: str

@router.post("/generate")
async def generate_report(request: ReportRequest):
    """生成学术报告（在已有的report_id目录中）"""
    try:
        user_id = get_current_user_id() or "anonymous"

        # 报告目录路径（模型结果已在此目录）
        report_dir = f"/root/librechat_user_data/{user_id}/reports/{request.report_id}"

        if not os.path.exists(report_dir):
            raise HTTPException(status_code=404, detail=f"报告目录不存在: {request.report_id}")

        # 读取模型结果（从report_id目录中）
        model_result_path = os.path.join(report_dir, "model_result.json")
        if not os.path.exists(model_result_path):
            raise HTTPException(status_code=404, detail="模型结果文件不存在")

        with open(model_result_path, 'r', encoding='utf-8') as f:
            model_result = json.load(f)

        # 使用ReportGenerator生成报告（传入report_dir）
        generator = ReportGenerator(report_dir)
        report_path = generator.generate_report(
            model_result=model_result,
            preprocessing_record=request.preprocessing_record,
            test_results=request.test_results,
            descriptive_stats=request.descriptive_stats,
            data_source_info=request.data_source_info,
            model_type=request.model_type
        )

        return {
            "success": True,
            "report_id": request.report_id,
            "report_path": report_path,
            "message": "报告生成成功"
        }

    except Exception as e:
        logger.error(f"报告生成失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/list")
async def list_reports():
    """获取用户所有报告列表"""
    try:
        user_id = get_current_user_id() or "anonymous"
        reports_base_dir = f"/root/librechat_user_data/{user_id}/reports"

        if not os.path.exists(reports_base_dir):
            return {"success": True, "reports": []}

        reports = []
        for report_id in os.listdir(reports_base_dir):
            report_dir = os.path.join(reports_base_dir, report_id)
            if not os.path.isdir(report_dir):
                continue

            meta_path = os.path.join(report_dir, "meta.json")
            if os.path.exists(meta_path):
                with open(meta_path, 'r', encoding='utf-8') as f:
                    meta = json.load(f)
                    # 过滤掉已删除的报告
                    if meta.get("status") != "deleted":
                        reports.append({
                            "report_id": report_id,
                            "report_name": meta.get("report_name", report_id),
                            "model_type": meta.get("model_type", "-"),
                            "data_source": meta.get("data_source", "-"),
                            "created_at": meta.get("created_at", 0),
                            "metrics": meta.get("metrics", {})
                        })

        # 按创建时间倒序排列
        reports.sort(key=lambda x: x["created_at"], reverse=True)

        return {
            "success": True,
            "reports": reports
        }

    except Exception as e:
        logger.error(f"获取报告列表失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/{report_id}")
async def delete_report(report_id: str):
    """软删除报告（标记为已删除，不实际删除文件）"""
    try:
        user_id = get_current_user_id() or "anonymous"
        report_dir = f"/root/librechat_user_data/{user_id}/reports/{report_id}"
        meta_path = os.path.join(report_dir, "meta.json")

        if not os.path.exists(meta_path):
            raise HTTPException(status_code=404, detail="报告不存在")

        # 读取meta.json
        with open(meta_path, 'r', encoding='utf-8') as f:
            meta = json.load(f)

        # 标记为已删除
        meta["status"] = "deleted"
        meta["deleted_at"] = int(__import__('time').time())

        # 保存更新后的meta.json
        with open(meta_path, 'w', encoding='utf-8') as f:
            json.dump(meta, f, ensure_ascii=False, indent=2)

        return {
            "success": True,
            "message": "报告已删除"
        }

    except Exception as e:
        logger.error(f"删除报告失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{report_id}/details")
async def get_report_details(report_id: str):
    """获取报告详细信息（包含完整模型结果）"""
    try:
        user_id = get_current_user_id() or "anonymous"
        report_dir = f"/root/librechat_user_data/{user_id}/reports/{report_id}"

        # 读取meta.json
        meta_path = os.path.join(report_dir, "meta.json")
        if not os.path.exists(meta_path):
            raise HTTPException(status_code=404, detail="报告不存在")

        with open(meta_path, 'r', encoding='utf-8') as f:
            meta = json.load(f)

        # 读取model_result.json
        model_result_path = os.path.join(report_dir, "model_result.json")
        model_result = None
        if os.path.exists(model_result_path):
            with open(model_result_path, 'r', encoding='utf-8') as f:
                model_result = json.load(f)

        return {
            "success": True,
            "report_id": report_id,
            "meta": meta,
            "model_result": model_result
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取报告详情失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/download/{report_id}/report.docx")
async def download_report(report_id: str):
    """下载报告DOCX文件"""
    try:
        user_id = get_current_user_id() or "anonymous"
        file_path = f"/root/librechat_user_data/{user_id}/reports/{report_id}/report.docx"

        if not os.path.exists(file_path):
            raise HTTPException(status_code=404, detail="报告文件不存在")

        return FileResponse(
            path=file_path,
            filename=f"{report_id}.docx",
            media_type='application/vnd.openxmlformats-officedocument.wordprocessingml.document'
        )

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
